File: Data Structures & Algorithms/count-connected-components/submission-7.py
class Union:
    def __init__(self,n):
        self.par = [i for i in range(n)]
        self.rank = [0] *n
    
    def getParent(self,n1):
        while self.par[n1] != n1:
            self.par[n1] = self.par[self.par[n1]]
            n1 = self.par[n1]
        return n1

    def union(self,n1,n2):
        p1,p2 = self.getParent(n1),self.getParent(n2)
        if p1 == p2:
            return False
        if self.rank[p1]>self.rank[p2]:
            self.par[p2] = p1
        elif self.rank[p2]>self.rank[p1]:
            self.par[p1] = p2
        else:
            self.rank[p1] +=1
            self.par[p2] = p1
        return True
    


class Solution:
    def countComponents(self, n: int, edges: List[List[int]]) -> int:
        u = Union(n)
        for e1,e2 in edges:
            if u.union(e1,e2):
                n-=1
        return n

        

File: Data Structures & Algorithms/redundant-connection/submission-7.java
class DSU {
    int[] parent,rank;
    public DSU(int n){
        parent = new int[n+1];
        rank = new int[n+1];
        for(int i = 0;i<=n;i++){
            parent[i] = i;
            rank[i] = 1;
        }

    }

    public int findParent(int e){
        int p = parent[e];
        while(p!=parent[p]){
            parent[p] = parent[parent[p]];
            p = parent[p];
        }
        return p;
    }

    public boolean union(int e1, int e2){
        var p1 = findParent(e1);
        var p2 = findParent(e2);
        if(p1==p2)
            return false;
        if(rank[p1]>rank[p2]){
            parent[p2] = p1;
            rank[p1] += rank[p2];
        } else {
            parent[p1] = p2;
            rank[p2] += rank[p1];
        }
        return true;
        
    }
}


class Solution {
    public int[] findRedundantConnection(int[][] edges) {
        DSU dsu = new DSU(edges.length);
        for(int[] edge:edges){
            int e1 = edge[0],e2 = edge[1];
            if (!dsu.union(e1,e2)){
                return edge;
            }
        }
        return null;
        
    }
}
